lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
-    // locations (a larger change).
-    if (v->getType().hasOpenedExistential())
-      return false;
-    args.emplace_back(v);
-  }
-  unsigned delta = params.size() - args.size();
-  assert(delta <= params.size()
-         && "Error, more args to partial apply than "
-            "params in its interface.");
-  params = params.drop_front(delta);
-
-  llvm::SmallVector<SILBasicBlock *, 2> exitingBlocks;
-  pai->getFunction()->findExitingBlocks(exitingBlocks);
-
-  // Go through our argument list and create new alloc_stacks for each
-  // non-trivial address value. This ensures that the memory location that we
-  // are cleaning up has the same live range as the partial_apply. Otherwise, we
-  // may be inserting destroy_addr of alloc_stack that have already been passed
-  // to a dealloc_stack.
-  for (unsigned i : llvm::reverse(indices(args))) {
-    SILValue arg = args[i];
-    SILParameterInfo paramInfo = params[i];
-
-    // If we are not going to destroy this partial_apply, continue.
-    if (!shouldDestroyPartialApplyCapturedArg(arg, paramInfo,
-                                              builder.getFunction()))
-      continue;
+  ArrayRef<SILParameterInfo> paramList = paiTy->getParameters();
+  auto argList = pai->getArgumentOperands();
+  paramList = paramList.drop_front(paramList.size() - argList.size());
+  SILFunction *F = pai->getFunction();
 
-    // If we have an object, we will not have live range issues, just continue.
-    if (arg->getType().isObject())
+  for (unsigned i : indices(argList)) {
+    Operand &argOp = argList[i];
+    SILParameterInfo param = paramList[i];
+    if (param.isIndirectMutating())
       continue;
 
-    // Now that we know that we have a non-argument address, perform a take-init
-    // of arg into a lifetime extended alloc_stack
-    args[i] = createLifetimeExtendedAllocStack(builder, loc, arg, exitingBlocks,
-                                               callbacks);
+    SILType ty = argOp.get()->getType();
+    if (!ty.isTrivial(*F) || (includeTrivialAddrArgs && ty.isAddress()))
+      argOperands.push_back(&argOp);
   }
+}
 
-  // Emit a destroy for each captured closure argument at each final release
-  // point.
-  for (auto *finalRelease : tracker.getFinalReleases()) {
-    builder.setInsertionPoint(finalRelease);
-    builder.setCurrentDebugScope(finalRelease->getDebugScope());
-    for (unsigned i : indices(args)) {
-      SILValue arg = args[i];
-      SILParameterInfo param = params[i];
-
-      releasePartialApplyCapturedArg(builder, loc, arg, param, callbacks);
+bool swift::collectDestroys(SingleValueInstruction *inst,
+                            SmallVectorImpl<SILInstruction *> &destroys) {
+  bool isDead = true;
+  for (Operand *use : inst->getUses()) {
+    SILInstruction *user = use->getUser();
+    if (useHasTransitiveOwnership(user)) {
+      if (!collectDestroys(cast<SingleValueInstruction>(user), destroys))
+        isDead = false;
+      destroys.push_back(user);
```

Comment:
Rather than else if, can you use early continues and a worklist instead of recursion?